server/src/main/java/org/elasticsearch/snapshots/RepositoryIntegrityHealthIndicatorService.java

```diff
@@ -73,78 +103,172 @@ public String name() {
+            if (size == 0) {
+                return null;
+            }
+
+            return String.format(Locale.ROOT, "[%d] %s snapshot repositor%s", size, type, size > 1 ? "ies" : "y");
+        }
+
+        public HealthIndicatorDetails getDetails(boolean verbose) {
+            if (verbose == false) {
+                return HealthIndicatorDetails.EMPTY;
+            }
+
+            if (healthStatus == GREEN) {
+                return new SimpleHealthIndicatorDetails(Map.of("total_repositories", totalRepositories));
+            }
+
+            return new SimpleHealthIndicatorDetails(
+                Map.of(
+                    "total_repositories",
+                    totalRepositories,
+                    "corrupted_repositories",
+                    corruptedRepositories.size(),
+                    "corrupted",
+                    limitSize(corruptedRepositories, 10),
+                    "unknown_repositories",
+                    unknownRepositoriesCount,
+                    "invalid_repositories",
+                    invalidRepositoriesCount
+                )
+            );
+        }
+
+        public List<HealthIndicatorImpact> getImpacts() {
+            if (healthStatus == GREEN) {
+                return List.of();
+            }
+            return IMPACTS;
+        }
+
+        public List<Diagnosis> getDiagnoses(int maxAffectedResourcesCount) {
+            var diagnoses = new ArrayList<Diagnosis>();
+            if (corruptedRepositories.isEmpty() == false) {
+                diagnoses.add(
+                    new Diagnosis(
+                        CORRUPTED_DEFINITION,
+                        List.of(new Diagnosis.Resource(SNAPSHOT_REPOSITORY, limitSize(corruptedRepositories, maxAffectedResourcesCount)))
+                    )
+                );
+            }
+            if (unknownRepositoriesCount > 0) {
+                diagnoses.add(createDiagnosis(UNKNOWN_DEFINITION, unknownRepositories, maxAffectedResourcesCount));
+            }
+            if (invalidRepositoriesCount > 0) {
+                diagnoses.add(createDiagnosis(INVALID_DEFINITION, invalidRepositories, maxAffectedResourcesCount));
+            }
+            return diagnoses;
+        }
+
+        private Diagnosis createDiagnosis(
+            Diagnosis.Definition definition,
+            Map<String, List<String>> reposMap,
+            int maxAffectedResourcesCount
+        ) {
+            // Flat map all the repo's (and make sure we only return distinct repo names).
+            var repoNames = reposMap.values().stream().flatMap(Collection::stream).distinct().limit(maxAffectedResourcesCount).toList();
+            var nodes = reposMap.entrySet()
+                .stream()
+                .filter(entry -> entry.getValue().isEmpty() == false)
+                .map(entry -> clusterState.nodes().get(entry.getKey()))
+                .limit(maxAffectedResourcesCount)
+                .toList();
```

Comment:
I think having different data structures here could benefit us. Also do not forget to sort the result to ensure our users have a stable view of the list.